src/memory/accessor.cpp

```diff
@@ -57,6 +57,7 @@ void accessor::increment(size_t value)
 
 accessor::~accessor()
 {
+    mutex_.unlock_upgrade_and_lock_shared();
```

Comment:
This results in an upgrade unlock in the case where assign has been called, yet the mutex is in a shared lock state at that point. This is a failure in the normal case (i.e. where assign is called). The presumption is that assign must be called once before destruct. The object is not initialized/usable otherwise. We could check the lock state on destruct and unlock shared or upgrade depending on that state, but the guard is unnecessary as the code will fail if the object is used incorrectly. It's preferred to encounter this failure based on the expectation that if assign is not called there is another error.